### PR TITLE
Bugs/2522 2554 2570/v8

### PR DESCRIPTION
**Pull request by victorjulien**
Link to redmine tickets:
https://redmine.openinfosecfoundation.org/issues/2574
https://redmine.openinfosecfoundation.org/issues/2573
https://redmine.openinfosecfoundation.org/issues/2570
https://redmine.openinfosecfoundation.org/issues/2554
https://redmine.openinfosecfoundation.org/issues/2522

Describe changes:

- small improvements of #3449: rewording, comments added
- #3448 (incl #3447)
- fixes for bugs 2522, 2554, 2570, 2573, 2574
- introduces a 'flush' logic for issue 2522, where to make sure stream and http bodies are inspected together, a 'stream flush' also triggers a 'http body' inspect: https://github.com/OISF/suricata/commit/7e004f52c60c5e4d7cd8f5ed09491291d18f42d2 and https://github.com/OISF/suricata/commit/35c5ae34589804768b75c9d04e715317f89c3447
- extend new json rule analyzer

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):
- PR victorjulien-pcap: https://buildbot.openinfosecfoundation.org/builders/victorjulien-pcap/builds/160
- PR victorjulien: https://buildbot.openinfosecfoundation.org/builders/victorjulien/builds/162

### CODE REVIEW

File: src/detect-engine-analyzer.c [hunk 1]
```diff
@@ -516,6 +516,74 @@ void EngineAnalysisRules2(const DetectEngineCtx *de_ctx, const Signature *s)
...
+        }
+        if (s->flags & SIG_FLAG_PDONLY) {
+            json_array_append_new(js_flags, json_string("proto_detect_only"));
+        }
+        if (s->flags & SIG_FLAG_SRC_IS_TARGET) {
+            json_array_append_new(js_flags, json_string("src_is_target"));
+        }
+        if (s->flags & SIG_FLAG_DEST_IS_TARGET) {
+            json_array_append_new(js_flags, json_string("dst_is_target"));
+        }
```
Comment (regit) [inline src/detect-engine-analyzer.c]: Should we add a comment somewhere to mention we need to a case here if ever we add a flag ? Or alternatively we could use a array of strings next to the flag declaration.

File: src/stream.h [hunk 2]
```diff
@@ -33,6 +33,7 @@
 #define STREAM_GAP              0x10    /**< data gap encountered */
 #define STREAM_DEPTH            0x20    /**< depth reached */
 #define STREAM_MIDSTREAM        0x40
+#define STREAM_FLUSH            0x80
```
Comment (regit) [inline src/stream.h]: BIT_U conversion maybe ?

File: src/detect.c [hunk 3]
```diff
@@ -984,6 +984,7 @@ static DetectRunScratchpad DetectRunSetup(
             if (p->proto == IPPROTO_TCP && pflow->protoctx &&
                     StreamReassembleRawHasDataReady(pflow->protoctx, p)) {
                 p->flags |= PKT_DETECT_HAS_STREAMDATA;
+                flow_flags |= STREAM_FLUSH;
```
Comment (regit) [inline src/detect.c]: Could we have some setup (small stream depth) where the stream_flush will cause the inspection to be done in the middle of a header ? Or are we sure we have post header data at that point ?

File: src/flow.c [hunk 4]
```diff
@@ -384,7 +384,12 @@ void FlowHandlePacketUpdate(Flow *f, Packet *p)
         }
     }
 
-    if ((f->flags & (FLOW_TO_DST_SEEN|FLOW_TO_SRC_SEEN)) == (FLOW_TO_DST_SEEN|FLOW_TO_SRC_SEEN)) {
+    if (SC_ATOMIC_GET(f->flow_state) == FLOW_STATE_ESTABLISHED) {
+        SCLogDebug("pkt %p FLOW_PKT_ESTABLISHED", p);
+        p->flowflags |= FLOW_PKT_ESTABLISHED;
+
```
Comment (regit) [inline src/flow.c]: white space

File: src/app-layer-parser.c [hunk 5]
```diff
@@ -130,7 +130,9 @@ typedef struct AppLayerParserProtoCtx_
     uint8_t first_data_dir;
 
     /* Option flags such as supporting gaps or not. */
-    uint64_t flags;
+    uint32_t option_flags;
```
Comment (regit) [inline src/app-layer-parser.c]: Maybe declare a cocinnelle check on flag here. This is getting messy to have 2 flags.

File: src/detect-engine.h [hunk 6]
```diff
@@ -102,6 +102,7 @@ int DetectEngineReloadIsIdle(void);
 int DetectEngineLoadTenantBlocking(uint32_t tenant_id, const char *yaml);
 int DetectEngineReloadTenantBlocking(uint32_t tenant_id, const char *yaml, int reload_cnt);
 
+int DetectEngineTentantRegisterLivedev(uint32_t tenant_id, int device_id);
```
Comment (regit) [inline src/detect-engine.h]: uint32_t and int in parameters of one single function catch my eyes.

File: src/detect-engine-analyzer.c (hunk 1 above)
Comment (victorjulien) [reply src/detect-engine-analyzer.c]: I've gone for the comments.

File: src/detect.c (hunk 3 above)
Comment (victorjulien) [reply src/detect.c]: Headers are not affected. Body data is reinspected until 'min_inspect_size' is reached.

File: src/flow.c (hunk 4 above)
Comment (victorjulien) [reply src/flow.c]: not seeing it, or do you mean the blank line? That one is deliberate

**Pull request merged**